architecture/src/androidLibMain/kotlin/observable/Observable.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ Copyright 2021 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+@file:JvmName("AndroidObservable")
+
+package com.splendo.kaluga.architecture.observable
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MediatorLiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.Observer
+import androidx.lifecycle.asLiveData
+import com.splendo.kaluga.architecture.observable.BaseSimpleDisposable
+import com.splendo.kaluga.architecture.observable.BasicSubject
+import com.splendo.kaluga.architecture.observable.Initialized
+import com.splendo.kaluga.architecture.observable.ObservableOptional
+import com.splendo.kaluga.architecture.observable.Uninitialized
+import com.splendo.kaluga.base.utils.EmptyCompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+import kotlin.coroutines.CoroutineContext
+import kotlin.properties.ReadOnlyProperty
+
+actual interface WithState<T> {
+    actual val stateFlow:StateFlow<T>
+    actual val valueDelegate: ReadOnlyProperty<Any?, T>
+    val liveData: LiveData<T>
+        get()= stateFlow.asLiveData()
+}
+
+val <R:T, T, OO: ObservableOptional<R>> BasicSubject<R, T, OO>.liveDataObserver: Observer<T>
+    get() = Observer<T> { this.post(it) }
+
+private fun <B, R:T, T, OO: ObservableOptional<R>> B.mutableLiveData(): MutableLiveData<R> where B : BasicSubject<R, T, OO>, B : WithMutableState<R> {
+    val mediatorLiveData = MediatorLiveData<R>()
+    mediatorLiveData.addSource(stateFlow.asLiveData()) { value ->
+        mediatorLiveData.postValue(value)
+    }
+
+    return mediatorLiveData
+}
+
+fun <T> LiveData<T>.observeOnCoroutine(coroutineScope: CoroutineScope, observer: Observer<T>) {
+    coroutineScope.launch(Dispatchers.Main.immediate) {
+        observeForever(observer)
+        val neverCompleting = EmptyCompletableDeferred()
+        neverCompleting.await()
+    }.invokeOnCompletion {
+        removeObserver(observer)
```

Comment:
I assume there are no tests because they are annoying to write for LiveData (though maybe it can still be attempted, perhaps as only on-device tests), but just want to check if this was well tested by hand?